![for sale image, from https://time.com/5835778/selling-home-coronavirus/](https://api.time.com/wp-content/uploads/2020/05/selling-home-coronavirus.jpg?w=800&quality=85)

# Project Title

## Overview

A one-paragraph overview of the project, including the business problem, data, methods, results and recommendations.

## Business Problem

Summary of the business problem you are trying to solve, and the data questions that you plan to answer to solve them.

Questions to consider:

- Who are your stakeholders? 

Real estate agency helping home buyers.

- What are your stakeholders' pain points related to this project?

- Why are your predictions important from a business perspective?

In [1]:
# Basic imports
import numpy as np
import pandas as pd
# Data visualizations
import matplotlib.pyplot as plt
import seaborn as sns
# Pre-Processing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
# Metrics
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

In [2]:
!ls ../../data

column_names.md
kc_house_data.csv


In [3]:
# To visualize the hidden columns in data
pd.options.display.max_columns=100 

## Data Understanding

Describe the data being used for this project.

Questions to consider:

- Where did the data come from, and how do they relate to the data analysis questions?
- What do the data represent? Who is in the sample and what variables are included?
- What is the target variable?
- What are the properties of the variables you intend to use?

In [4]:
# code here to explore your data
df = pd.read_csv("../../data/kc_house_data.csv")

In [5]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors waterfront  view  condition          grade  sqft_above  \
0      5650     1.0        NaN  NONE    Average      7 Average        1180   
1      7242     2.0         NO  NONE    Average      7 Average        2170   
2     10000     1.0         NO  NONE    Average  6 Low Average         770   
3      5000     1.0         NO  NONE  Very Good      7 Average        1050   
4      8080     1.0         NO  NONE    Average         8 Good        1680   

  sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0           0.0      1955           0.0    98178  47.5112 -122.257   
1         400.0      1951        1991.0    98125  47.7210 -122.319   
2           0.0      1933           NaN    98028  47.7379 -122.233   
3         910.0      1965           0.0    98136  47.5208 -122.393   
4           0.0      1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503

In [6]:
df.shape

(21597, 21)

In [7]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    sqft_above      yr_built  yr_renovated  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  17755.000000   
mean   1.509941e+04      1.494096   1788.596842   1970.999676     83.636778   
std    4.141264e+04      0.539683    827.759761     29.375234    399.946414   
min    5.200000e+02      1.000000    370.000000   1900.000000      0.000000   
25%    5.040000e+03      1.000000   1190.000000   1951.000000      0.000000   
50%    7.618000e+03      1.500000   1560.000000   1975.000000      0.000000   
75%    1.068500e+04      2.000000   2210.000000   1997.000000      0.000000   
max    1.651359e+06      3.500000   9410.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000  21597.000000   21597.000000   21597.000000  
mean   98077.951845     47.560093   -122.213982    1986.620318   12758.283512  
std       53.513072      0.138552      0.140724     685.230472   27274.441950  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471100   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.231000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

Price is the target variable for our business problem.

The data includes Kings County houses' information and sell prices from 1900 to 2015.

In [9]:
#sns.pairplot(df)

## Data Preparation

Describe and justify the process for preparing the data for analysis.

Questions to consider:

- Were there variables you dropped or created?
- How did you address missing values or outliers?
- Why are these choices appropriate given the data and the business problem?

In [10]:
df['id'].value_counts()

795000620     3
1825069031    2
2019200220    2
7129304540    2
1781500435    2
             ..
7812801125    1
4364700875    1
3021059276    1
880000205     1
1777500160    1
Name: id, Length: 21420, dtype: int64

In [11]:
df[df['id']==795000620] # Multiple rows for same house means that the house is sold multiple times!

id        date     price  bedrooms  bathrooms  sqft_living  \
17588  795000620   9/24/2014  115000.0         3        1.0         1080   
17589  795000620  12/15/2014  124000.0         3        1.0         1080   
17590  795000620   3/11/2015  157000.0         3        1.0         1080   

       sqft_lot  floors waterfront  view condition   grade  sqft_above  \
17588      6250     1.0         NO  NONE      Fair  5 Fair        1080   
17589      6250     1.0         NO  NONE      Fair  5 Fair        1080   
17590      6250     1.0        NaN  NONE      Fair  5 Fair        1080   

      sqft_basement  yr_built  yr_renovated  zipcode      lat    long  \
17588           0.0      1950           0.0    98168  47.5045 -122.33   
17589           0.0      1950           0.0    98168  47.5045 -122.33   
17590           0.0      1950           NaN    98168  47.5045 -122.33   

       sqft_living15  sqft_lot15  
17588           1070        6250  
17589           1070        6250  
17590           1070        6250

We should keep the multiple rows to reflect price change over the years. 

In [12]:
df['bedrooms'].value_counts()

3     9824
4     6882
2     2760
5     1601
6      272
1      196
7       38
8       13
9        6
10       3
11       1
33       1
Name: bedrooms, dtype: int64

In [13]:
df[df['bedrooms']==33]

id       date     price  bedrooms  bathrooms  sqft_living  \
15856  2402100895  6/25/2014  640000.0        33       1.75         1620   

       sqft_lot  floors waterfront  view  condition      grade  sqft_above  \
15856      6000     1.0         NO  NONE  Very Good  7 Average        1040   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
15856         580.0      1947           0.0    98103  47.6878 -122.331   

       sqft_living15  sqft_lot15  
15856           1330        4700

It is impossible to have 33 bedrooms house with 1620 sqrt living area and 1 floor. There must be typing error. We should replace 33 with the mean of number of bedrooms, which is 3.

In [14]:
df['bedrooms'].replace(33, 3, inplace=True)

In [15]:
df['bedrooms'].value_counts()

3     9825
4     6882
2     2760
5     1601
6      272
1      196
7       38
8       13
9        6
10       3
11       1
Name: bedrooms, dtype: int64

In [16]:
df[df['bedrooms']==11]

id       date     price  bedrooms  bathrooms  sqft_living  \
8748  1773100755  8/21/2014  520000.0        11        3.0         3000   

      sqft_lot  floors waterfront  view condition      grade  sqft_above  \
8748      4960     2.0         NO  NONE   Average  7 Average        2400   

     sqft_basement  yr_built  yr_renovated  zipcode     lat     long  \
8748         600.0      1918        1999.0    98106  47.556 -122.363   

      sqft_living15  sqft_lot15  
8748           1420        4960

In [17]:
df[df['bedrooms']==10]

id        date      price  bedrooms  bathrooms  sqft_living  \
13301   627300145   8/14/2014  1150000.0        10       5.25         4590   
15147  5566100170  10/29/2014   650000.0        10       2.00         3610   
19239  8812401450  12/29/2014   660000.0        10       3.00         2920   

       sqft_lot  floors waterfront     view condition      grade  sqft_above  \
13301     10920     1.0         NO  AVERAGE   Average   9 Better        2500   
15147     11914     2.0         NO     NONE      Good  7 Average        3010   
19239      3745     2.0         NO     NONE      Good  7 Average        1860   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
13301        2090.0      2008           0.0    98004  47.5861 -122.113   
15147         600.0      1958           0.0    98006  47.5705 -122.175   
19239        1060.0      1913           0.0    98105  47.6635 -122.320   

       sqft_living15  sqft_lot15  
13301           2730       10400  
15147           2040       11914  
19239           1810        3745

Houses with 11 and 10 bedrooms looks okay for the analysis.

In [18]:
df['zipcode'].value_counts()

98103    602
98038    589
98115    583
98052    574
98117    553
        ... 
98102    104
98010    100
98024     80
98148     57
98039     50
Name: zipcode, Length: 70, dtype: int64

In [19]:
df['yr_built'].value_counts()

2014    559
2006    453
2005    450
2004    433
2003    420
       ... 
1933     30
1901     29
1902     27
1935     24
1934     21
Name: yr_built, Length: 116, dtype: int64

#### Check null values in the dataframe

In [20]:
df.isna().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

The columns named as waterfront, view, and yr_renovated have null values.

In [21]:
# waterfront indicates Whether the house is on a waterfront
df['waterfront'].value_counts()

NO     19075
YES      146
Name: waterfront, dtype: int64

In [22]:
df[(df['waterfront']!= 'YES') & (df['waterfront']!= 'NO')]

id        date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520  10/13/2014  221900.0         3       1.00         1180   
10     1736800520    4/3/2015  662500.0         3       2.50         3560   
23     8091400200   5/16/2014  252700.0         2       1.50         1070   
40     5547700270   7/15/2014  625000.0         4       2.50         2570   
55     9822700295   5/12/2014  885000.0         4       2.50         2830   
...           ...         ...       ...       ...        ...          ...   
21578  5087900040  10/17/2014  350000.0         4       2.75         2500   
21582  8956200760  10/13/2014  541800.0         4       2.50         3118   
21586   844000965   6/26/2014  224000.0         3       1.75         1500   
21587  7852140040   8/25/2014  507250.0         3       2.50         2270   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   

       sqft_lot  floors waterfront     view condition          grade  \
0          5650     1.0        NaN     NONE   Average      7 Average   
10         9796     1.0        NaN     NONE   Average         8 Good   
23         9643     1.0        NaN     NONE   Average      7 Average   
40         5520     2.0        NaN     NONE   Average       9 Better   
55         5000     2.0        NaN     NONE   Average       9 Better   
...         ...     ...        ...      ...       ...            ...   
21578      5995     2.0        NaN     NONE   Average         8 Good   
21582      7866     2.0        NaN  AVERAGE   Average       9 Better   
21586     11968     1.0        NaN     NONE   Average  6 Low Average   
21587      5536     2.0        NaN     NONE   Average         8 Good   
21595      2388     2.0        NaN     NONE   Average         8 Good   

       sqft_above sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180           0.0      1955           0.0    98178  47.5112   
10           1860        1700.0      1965           0.0    98007  47.6007   
23           1070           0.0      1985           NaN    98030  47.3533   
40           2570           0.0      2000           NaN    98074  47.6145   
55           2830           0.0      1995           0.0    98105  47.6597   
...           ...           ...       ...           ...      ...      ...   
21578        2500           0.0      2008           0.0    98042  47.3749   
21582        3118           0.0      2014           0.0    98001  47.2931   
21586        1500           0.0      2014           0.0    98010  47.3095   
21587        2270           0.0      2003           0.0    98065  47.5389   
21595        1600           0.0      2004           0.0    98027  47.5345   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
10    -122.145           2210        8925  
23    -122.166           1220        8386  
40    -122.027           2470        5669  
55    -122.290           1950        5000  
...        ...            ...         ...  
21578 -122.107           2530        5988  
21582 -122.264           2673        6500  
21586 -122.002           1320       11303  
21587 -121.881           2270        5731  
21595 -122.069           1410        1287  

[2376 rows x 21 columns]

Let's convert YES value as 1, NO value as 0, and also replace NaN values with 0:

In [23]:
df['waterfront'].replace("YES", 1, inplace=True)

In [24]:
df['waterfront'].replace("NO", 0, inplace=True)

In [25]:
df['waterfront'].fillna(value=0,inplace=True)

In [26]:
df['waterfront'].value_counts()

0.0    21451
1.0      146
Name: waterfront, dtype: int64

In [27]:
df.isna().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront          0
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

In [28]:
df['view'].value_counts()

NONE         19422
AVERAGE        957
GOOD           508
FAIR           330
EXCELLENT      317
Name: view, dtype: int64

In [42]:
df.dropna(subset=['view'], inplace=True)

We dropped 63 rows of the data which have missing view value.

In [30]:
df['yr_renovated'].value_counts()

0.0       17011
2014.0       73
2003.0       31
2013.0       31
2007.0       30
          ...  
1946.0        1
1959.0        1
1971.0        1
1951.0        1
1954.0        1
Name: yr_renovated, Length: 70, dtype: int64

Most of the houses are not renovated, so we can replace NaN values with zero.

In [31]:
df['yr_renovated'].fillna(value=0,inplace=True)

In [43]:
df.isna().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

We cleaned all of the missing values from our dataframe.

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     21597 non-null  float64
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   21597 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

The columns named date, view, condition, grade, and sqrt_basement are stored as objects, not numerical values.

In [51]:
# We will look this after changing objects to numerics.
np.abs(df.corr()['price']).sort_values()

id               0.017392
long             0.022417
zipcode          0.053620
yr_built         0.054273
sqft_lot15       0.083189
sqft_lot         0.090338
yr_renovated     0.117982
floors           0.257052
waterfront       0.259220
lat              0.307868
bedrooms         0.315229
bathrooms        0.525053
sqft_living15    0.585304
sqft_above       0.605695
sqft_living      0.701587
price            1.000000
Name: price, dtype: float64

## Modeling

Describe and justify the process for analyzing or modeling the data.

Questions to consider:

- How did you analyze the data to arrive at an initial approach?
- How did you iterate on your initial approach to make it better?
- Why are these choices appropriate given the data and the business problem?

## Evaluation

The evaluation of each model should accompany the creation of each model, and you should be sure to evaluate your models consistently.

Evaluate how well your work solves the stated business problem. 

Questions to consider:

- How do you interpret the results?
- How well does your model fit your data? How much better is this than your baseline model? Is it over or under fit?
- How well does your model/data fit any modeling assumptions?

For the final model, you might also consider:

- How confident are you that your results would generalize beyond the data you have?
- How confident are you that this model would benefit the business if put into use?

### Baseline Understanding

- What does a baseline, model-less prediction look like?

In [34]:
# code here to arrive at a baseline prediction

### First $&(@# Model

Before going too far down the data preparation rabbit hole, be sure to check your work against a first 'substandard' model! What is the easiest way for you to find out how hard your problem is?

In [35]:
# code here for your first 'substandard' model

In [36]:
# code here to evaluate your first 'substandard' model

### Modeling Iterations

Now you can start to use the results of your first model to iterate - there are many options!

In [37]:
# code here to iteratively improve your models

In [38]:
# code here to evaluate your iterations

### 'Final' Model

In the end, you'll arrive at a 'final' model - aka the one you'll use to make your recommendations/conclusions. This likely blends any group work. It might not be the one with the highest scores, but instead might be considered 'final' or 'best' for other reasons.

In [39]:
# code here to show your final model

In [40]:
# code here to evaluate your final model

## Conclusions

Provide your conclusions about the work you've done, including any limitations or next steps.

Questions to consider:

- What would you recommend the business do as a result of this work?
- What are some reasons why your analysis might not fully solve the business problem?
- What else could you do in the future to improve this project (future work)?
